# Webscraping Tutorial


This notebook will attempt to scrape the corporate information from a list of selected companies. This was done as a favor for a friend.


# Importing Libraries and Text File


In [1]:
import pandas as pd
from bs4 import BeautifulSoup
import numpy as np
import requests
from tqdm import tqdm_notebook


First, I will look at the URL list from the text file and import it into python as a list


In [2]:
# Import URL text file into list
url_list = []
url = open("sgx_links.txt", 'r', newline=None)
for line in url:
    url_list.append(line)


In [3]:
# Remove \n from the each URL
url_list = [i.replace('\n', '') for i in url_list]


In [5]:
len(url_list)


707

Based on the text file, there are 707 URLs in the list.

# Webscraping 

First, I will create a dataframe to hold the information scraped.

In [4]:
info = pd.DataFrame(columns=['company_name', 'incorporated', 'incorporated_on', 'ISIN_code', 'registered_office', 'telephone', 'fax', 'email',
                    'secretary', 'website', 'listing', 'listing_board', 'other_stock_exchange', 'registrars', 'registrars_address', 'auditors', 'background'])


As I will be scraping directly from the SGX website, I will be using Beautiful Soup HTML parser to scrape the information that are needed.

In [7]:
#Webscrape with BeautifulSoup

info_list = []

for url in tqdm_notebook(url_list):

    URL = url
    page = requests.get(URL)
    soup = BeautifulSoup(page.content, "html.parser")

    company_name = soup.find(id="ctl07_compFullNameLabel").text
    incorporated = soup.find(id="ctl07_incorporatedLabel").text
    incorporated_on = soup.find(id='ctl07_incorpOnLabel').text
    isin_code = soup.find(id='ctl07_isinCodeLabel').text
    registered_office = soup.find(
        'dt', text='Registered Office :').findNext("dd").text
    telephone = soup.find('dt', text='Telephone :').findNext("dd").text
    fax = soup.find('dt', text='Fax :').findNext("dd").text
    email = soup.find('dt', text='Email :').findNext("dd").text
    secretary = soup.find('dt', text='Secretary :').findNext("dd").text
    website = soup.find(
        'dt', text='Link to Internet Website :').findNext("dd").text
    listing = soup.find(id='ctl07_listingDateLabel').text
    listing_board = soup.find(id='ctl07_lbllistingBoard').text

    # Not all url has this row so the webscrape will try to scrape it and return NAN if none
    try:
        other_stock_exchange = soup.find(
            text='OTHER STOCK EXCHANGE LISTINGS').findNext('dd').text
    except (AttributeError):
        other_stock_exchange = np.nan

    registrars = soup.find(
        text='REGISTRARS / TRANSFER AGENTS & ADDRESS').findNext('dd').text
    registrars_address = soup.find(
        text='REGISTRARS / TRANSFER AGENTS & ADDRESS').findNext('dd').findNext('dd').text
    auditors = soup.find(text='AUDITORS').findNext('dd').text
    background = soup.find(id='litIPOCompany').text

    info_list.append([company_name, incorporated, incorporated_on, isin_code, registered_office, telephone, fax, email,
                     secretary, website, listing, listing_board, other_stock_exchange, registrars, registrars_address, auditors, background])


/var/folders/sz/qlhtygjx661flzpy8w5byj240000gn/T/ipykernel_26933/2163992587.py:3: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for url in tqdm_notebook(url_list):


  0%|          | 0/707 [00:00<?, ?it/s]

In [12]:
#Append it to the pandas dataframe
info = info.append(pd.DataFrame(info_list, columns=info.columns))


/var/folders/sz/qlhtygjx661flzpy8w5byj240000gn/T/ipykernel_26933/4238377209.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  info = info.append(pd.DataFrame(info_list, columns = info.columns))


Finally, I will need to clean the data to remove all the \r and \n markings

In [18]:
info = info.replace('\n', '', regex=True)
info = info.replace('\r', '', regex=True)


In [24]:
info


company_name  \
0                                      3CNERGY LIMITED   
1                                 5E RESOURCES LIMITED   
2                        8TELECOM INTL HOLDINGS CO LTD   
3                                           9R LIMITED   
4                           ABACUS CAPITAL (S) PTE LTD   
..                                                 ...   
702                  YUUZOO NETWORKS GROUP CORPORATION   
703  ZHENENG JINJIANG ENVIRONMENT HOLDING COMPANY L...   
704                  ZHONGMIN BAIHUI RETAIL GROUP LTD.   
705                   ZHONGXIN FRUIT AND JUICE LIMITED   
706                                 ZICO HOLDINGS INC.   

                incorporated incorporated_on     ISIN_code  \
0                  SINGAPORE     24 Feb 1973  SG0502000029   
1                  SINGAPORE     18 Oct 2021  SGXE78399073   
2                    BERMUDA     05 Jan 2004  BMG3087Y2074   
3                  SINGAPORE     04 Nov 1993  SGXE45420721   
4                  SINGAPORE                  SG1BE4000002   
..                       ...             ...           ...   
702  VIRGIN ISLANDS, BRITISH     22 Mar 2005  VGG9889R1001   
703           CAYMAN ISLANDS     08 Sep 2010  KYG9898S1075   
704                SINGAPORE     17 Sep 2004  SG2C76966531   
705                SINGAPORE     27 Sep 2002  SG1P25916898   
706                 MALAYSIA     09 Dec 2010  MYA007968003   

                                     registered_office  \
0    82 Ubi Avenue 4#05-04 Edward Boustead CentreSi...   
1    30 Cecil Street#19-08Prudential TowerSingapore...   
2    Clarendon House2 Church StreetHamilton HM 11Be...   
3    105 Cecil Street,#12-02 The Octagon,Singapore ...   
4                                                  NA    
..                                                 ...   
702  Jayla PlaceWickhams Cay 1Road TownTortola Brit...   
703  Grand Pavillion, Hibiscus Way802 West Bay Road...   
704  160 Robinson RoadSBF Centre#15-06Singapore 068...   
705  No 25 International Business Park#02-53 German...   
706  Unit Level 13(A), Main Office Tower,Financial ...   

                           telephone                             fax  \
0                       65 69708117                                    
1                                                                      
2                  (86) 57188225288                 (86) 57188225291   
3                      65 6601 9500                     65 6601 9600   
4                                                                      
..                               ...                             ...   
702                     65 62713468                      65 62758469   
703                 86 571 87699700                 86 571 88388848.   
704  86 592 5863888 (Xiamen) 65 644   86 592 5182791 (Xiamen) 65 644   
705                     65 65572308                                    
706                  (65) 6438 7929                                    

                        email                        secretary  \
0    enquiries@3cnergy.com.sg                    Cheok Hui Yee   
1    finance@5e-resources.com              Tan Sey Liy Shirley   
2                                                                
3                                        Lai Kuan Loong Victor   
4                                                                
..                        ...                              ...   
702                                                              
703                            Toh Li Ping AngelaHoon Chi Tern   
704                                             Chia Foon Yeow   
705                                             Lee Wei Hsiung   
706                                   ZICO Secretarial Limited   

                              website  \
0          http://www.3cnergy.com.sg/   
1         http://www.5e-resources.com   
2              http://www.8telecom.cn   
3             http://www.vikingom.com   
4                                       

Finally, I will be exporting the information to a CSV file

In [22]:
info.to_csv('data_cleaned.csv')
